Makefile

```diff
@@ -32,7 +33,14 @@ test-unit: venv
 	venv/bin/tox -e $(TOX_ENVLIST) -- $(TOX_ARGS)
 
 test-integration: venv
-	CACHI2_GENERATE_TEST_DATA=$(GENERATE_TEST_DATA) venv/bin/tox -e integration
+	if [ $(TEST_LOCAL_PYPISERVER) = true ]; then \
+		PYPISERVER_PORT=$(PYPISERVER_PORT) PYPISERVER_IMAGE=$(PYPISERVER_IMAGE) tests/pypiserver/start.sh & \
+		trap "kill $$!" EXIT && \
+		while true; do sleep 1; curl -fsS http://localhost:$(PYPISERVER_PORT) && break; done; \
+	fi && \
```

Comment:
I am very much not a fan of abusing the Makefile for this purpose (particularly to use Bash as a crutch instead of modeling this properly in the Python test suite). It really only should be a recipe for tedious build/test/release processes. Now that said, one might argue that this indeed is related to the "test" part I mentioned in the previous sentence, but not quite - configuring tox is a good example of that. If we optionally need a test server deployed for a particular test's needs then I believe we should do it part of the test suite and not part of the Makefile.
Related to your `integration/conftest.py` changes, is there anything preventing us to model this server as a test fixture with a cleanup section based on Python's subprocess? After all, you guard the test with a pytest  `skip` mark, so such a fixture should not run in that case and it would be nicely contained within the test suite where it belongs.

Bonus points if we could reuse the `utils.ContainerImage` class to do this which I think it should be possible looking at the command line the container gets.

Reply:
I can try that, but I'd like to keep the `pypiserver-run` makefile target as well

The process will need to run in the background, not sure how pretty that will be

Reply:
> The process will need to run in the background, not sure how pretty that will be

I'm not sure I follow, `subprocess.Popen` doesn't block the main process, IIUC if you put the Popen call to a fixture followed by a `yield` and then for the "cleanup" kill and wait for the server process that should theoretically work, shouldn't it?

Reply:
> pytest eats the stdout and stderr from the subprocess

Although with `--capture=sys` or `--capture=tee-sys` passed to pytest, this looks better. Still need to find out what those do though

Reply:
Sigh, true. I guess if only

>  if the server fails to start, you'll never find out why (the output is logged when the tests finish

needs to be solved, I think redirecting podman's stdout to a file instead of the parent's process stdout (which pytest consumes) then that should also work, but then

> but it's already a bit worse than bash

will still remain true I suppose. Can we at least extract the Makefile bits, stuff it in a bash script and bury it somewhere under the integration test directory hierarchy? That shouldn't have an effect on the github action you needed the dedicated makefile target for.

Reply:
> Although with --capture=sys or --capture=tee-sys passed to pytest, this looks better. Still need to find out what those do though

Looking at the docs `tee-sys` reads exactly like what we need.

Reply:
Added a WIP commit for what it would look like, see if you like it better

Reply:
> Added a WIP commit for what it would look like, see if you like it better

Unsurprisingly, it led to more lines in Python, but the code is still perfectly readable and contained within the test suite, nicely done!